packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts

```diff
@@ -0,0 +1,242 @@
+import { IExecuteFunctions } from 'n8n-core';
+import {
+	IDataObject,
+	INodeExecutionData,
+	INodeType,
+	INodeTypeDescription,
+} from 'n8n-workflow';
+
+import {
+	googleApiRequest,
+	encodeEmail,
+} from './GenericFunctions';
+
+import {
+	emailOperations,
+	emailFields
+} from './EmailDescriptions';
+
+import {
+	labelOperations,
+	labelFields,
+} from './LabelDescriptions';
+
+import * as _ from 'lodash';
+
+export class Gmail implements INodeType {
+	description: INodeTypeDescription = {
+		displayName: 'Gmail',
+		name: 'gmail',
+		icon: 'file:gmail.png',
+		group: ['transform'],
+		version: 1,
+		subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
+		description: 'View, Write, and Send Emails and Manage Your Inbox',
+		defaults: {
+			name: 'Gmail',
+			color: '#772244',
+		},
+		inputs: ['main'],
+		outputs: ['main'],
+		credentials: [
+			{
+				name: 'gmailOAuth2',
+				required: true,
+			},
+		],
+		properties: [
+			{
+				displayName: 'Resource',
+				name: 'resource',
+				type: 'options',
+				options: [
+					{
+						name: 'Labels',
+						value: 'label',
+					},
+					{
+						name: 'Emails',
+						value: 'email',
+					},
+				],
+				default: 'label',
+				description: 'The resource to operate on.',
+			},
+			...labelOperations,
+			...labelFields,
+			...emailOperations,
+			...emailFields
+		]
+	};
+
+	async execute(this: IExecuteFunctions): Promise<INodeExecutionData[][]> {
+		const items = this.getInputData();
+
+		const resource = this.getNodeParameter('resource', 0) as string;
+		const operation = this.getNodeParameter('operation', 0) as string;
+
+		let method = '';
+		let body: IDataObject = {};
+		let qs: IDataObject = {};
+		let endpoint = '';
+
+		const email =  {
+			from: '',
+			to: '',
+			cc: '',
+			bcc: '',
+			inReplyTo: '',
+			reference: '',
+			subject: '',
+			body: '',
+			attachments: []
+		};
+
+		for (let i = 0; i < items.length; i++) {
+			if(resource === 'label') {
+				method = 'POST';
+
+				if(operation === 'create') {
+					// https://developers.google.com/gmail/api/v1/reference/users/labels/create
+					endpoint = '/gmail/v1/users/me/labels';
+
+					const labelName = this.getNodeParameter('name', i) as string;
+
+					body = {
+						'labelListVisibility': 'labelShow',
+						'messageListVisibility': 'show',
+						'name': labelName
+					};
+				} else if(['add', 'delete'].includes(operation)) {
+					// https://developers.google.com/gmail/api/v1/reference/users/messages/modify
+					const messageID = this.getNodeParameter('messageID', i);
+					const labelID = this.getNodeParameter('labelID', 0);
```

Comment:
I have not checked the logic yet, but `0` instead of `i` _might_ not be right here. Please double check.